### ISSUE TITLE
Major refactoring

### ISSUE THREAD

**Issue by deltaidea** [suclearnub/budgetsim#7]
The main issue with current code base is that it's all in one big file. And it's written by two C++ noobs.

This issue is the place for meta refactoring discussion, meaning that it's for general questions, manuals, etc. Separate issues should be created for individual refactoring solutions.

**Comment by deltaidea** [comment]
Ideally, there should be a main file which `#include`s all the implementation details. It should look similar to this:

``` C++
int main() {
    Game game = new Game();
    game.start();
    while (!game.isEnded) {
        game.step();
    }
}
```

**Comment by deltaidea** [comment]
Also ideally, there should be at least unit tests.

I'll read about [TDD](https://en.wikipedia.org/wiki/Test-driven_development) in C++ later.
And if we'll use TDD, it's better to just rewrite everything from scratch.

**Comment by seanballais** [comment]
+1 for the game loop! :)

**Comment by deltaidea** [comment]
@seanballais, how do you TDD in C++?

I found [biicode](http://biicode.com/) as a dependency manager and [Catch](https://github.com/philsquared/Catch) and [Google Test](https://code.google.com/p/googletest/) as unit testing tools.

Is that how you do stuff in C++?

My main background is JS and a little bit of Python, that may explain why I search for a package/dependency manager.

The thought process here is:
- It's easier to write code when you have tests.
- To run tests, you need a test runner (duh).
- Test runner is a third-party dependency. I don't want to write my own.
- Manual installation of third-party stuff leads to dependency hell.

**Comment by the-emerald** [comment]
Should some features be put into functions first? Just a thought.

**Comment by deltaidea** [comment]
Yes. That's the whole point. Tests just help you do it easier and better.

**Comment by seanballais** [comment]
@deltaidea Sorry for the late reply. I think you can use cppUnit for testing. Still looking into Google Test, biicode, and Catch.
@suclearnub How about placing all your game tasks in a text file instead? If you do this, you can say hello to modders.

**Comment by the-emerald** [comment]
@seanballais Ah, so documentation? I'll work on it right away.

**Comment by seanballais** [comment]
@suclearnub That's not what I meant. I mean put

```
 actionChoices = {    
        {    
            "Improve the toilets ($20,000).",    
        "You've improved the toilets.",    
        [&]() {    
            schoolBudget -= 20000;    
            reputation += 0.4;    
        }    
    },    
    {    
        "Put flower pots for more beauty ($15,000).",
        "You put flower pots around the school.",     
        [&]() {    
            schoolBudget -= 15000;    
            reputation += 0.1;    
        }    
    },    
  ... // some other part of actionChoices    
```

 to a separate text file, or maybe a JSON file.

**Comment by deltaidea** [comment]
@seanballais I don't think it's possible. These tasks have custom functions in them. You can't put them into JSON. Evaluating that code at runtime from strings, if at all possible, is a bad idea. We can use some kind of [DSL](https://en.wikipedia.org/wiki/Domain-specific_language), but is it worth it?..

**Comment by seanballais** [comment]
@deltaidea I don't think it's worth it. We can just put the actionChoices in a JSON/text file with this format.

``` js
{
    "Put flower pots for more beauty ($15, 000)", // a possible action
    "You placed flower pots around the school.", // what will be printed
    -15000, // the amount the budget will be reduced or added
    0.1 // the reputation reduced or added
}
```

**Comment by deltaidea** [comment]
@seanballais Yeah, but there're only two actions like that, dealing only with those two variables. Then you have `rand()`, other variables, reading from stdin, loops, ... You can't serialize/stringify that, which is unfortunate.

**Comment by seanballais** [comment]
@deltaidea We can use [nlohmann's](https://github.com/nlohmann/) [JSON for Modern C++ project](https://github.com/nlohmann/json) to parse JSON.

**Comment by deltaidea** [comment]
I get it, of course there're libraries for JSON in C++.

What I'm talking about is that we can't serialize, for example, this code:

``` C++
{
    "Introduction to school ($5000). Chance of increasing or decreasing population.",
    "You introduce more people to the school.",
    [&]() {
        schoolBudget -= 5000;
        cout << "The event went ";
        int rng = rand() % 4;
        if(rng>=2) {
            rng = 10;
            cout << "alright. You gained 10 students! \n";
        }
        else {
            rng = -10;
            cout << "terribly wrong. Students are leaving! You've lost 10 students.\n";
        }
    }
},
```

**Comment by seanballais** [comment]
Ah that! Well, we just think of a different format. But we should really separate data from code, if possible.

**Comment by deltaidea** [comment]
Yes.

First, let's refactor all of the school code out of `main.cpp` into a `School` class: #10. Then it will be easier to split it further.

**Comment by seanballais** [comment]
Great!

Don't forget to create tests whenever you add additional functionality. Then we also create a `Game` class.
